src/corelibs/U2Core/src/io/IOAdapterTextStream.h

```diff
@@ -0,0 +1,155 @@
+class U2OpStatus;
+
+/**
+ * IO Adapter device used by text streams.
+ * Notes:
+ *  The supported feature set was tested to work only with QTextStream.
+ *  See original QT docs for the methods with 'override'.
+ *
+ */
+class IOAdapterDevice : public QIODevice {
+public:
+    /**
+     * Creates a new unbuffered QIODevice over IOAdapter instance.
+     * The 'ioAdapter' must be in 'open' state.
+     * The QIODevice is writable only if the  'ioAdapter' is writable.
+     */
+    IOAdapterDevice(IOAdapter *ioAdapter, QObject *parent = nullptr);
+
+    bool atEnd() const override;
+
+    qint64 pos() const override;
+
+    bool seek(qint64 seekPos) override;
+
+protected:
+    qint64 readData(char *data, qint64 length) override;
+
+    qint64 writeData(const char *data, qint64 length) override;
+
+private:
+    IOAdapter *ioAdapter;
+    qint64 currentPos;
+};
+
+/** Base class for reader & writer io-adapter text streams. */
+class U2CORE_EXPORT IOAdapterReaderAndWriterBase {
+public:
+    /** Initializes underlying qt stream & io-device. If 'codec' is null uses default locale codec. */
+    IOAdapterReaderAndWriterBase(IOAdapter *ioAdapter, QTextCodec *codec = nullptr);
+
+    /** Returns url from the ioAdapter. */
+    GUrl getURL() const;
+
+    /** Returns factory from the ioAdapter. */
+    IOAdapterFactory *getFactory() const;
+
+protected:
+    IOAdapter *ioAdapter;
+    QScopedPointer<QIODevice> ioDevice;
+    QTextStream stream;
+};
+
+/**
+ * Text stream reader over the opened ioAdapter instance.
+ * Automatically detects the encoding and allows reading of the binary data as a text.
+ * If the codec can't be detected from the binary data falls back to the most suitable codec for the current user locale (usually UTF-8).
+ */
+class U2CORE_EXPORT IOAdapterReader : public IOAdapterReaderAndWriterBase {
+public:
+    /**
+     * Initializes text stream with the ioAdapter wrapped into QIODevice.
+     * Performs no actual reading from the ioAdapter.
+     */
+    IOAdapterReader(IOAdapter *ioAdapter);
+
+    /** Pushes back all unused buffers to ioAdapter by calling 'skip' method. */
+    ~IOAdapterReader();
+
+    /**
+     * Reads the stream until one of the separators is found. Saves the result to 'result' string buffer.
+     * Clears the 'result' buffer first before reading.
+     * Returns number of characters read, same as result.length().
+     */
+    int read(U2OpStatus &os, QString &result, int maxLength, const QBitArray &terminators, IOAdapter::TerminatorHandling terminatorMode, bool *terminatorFound = nullptr);
+
+    /*
+     * Returns a progress value in the range 0..100, or a negative value if the progress is unknown/not supported
+     * Proxies the call to ioAdapter.getProgress method.
+     */
+    int getProgress() const;
+
+    /** Undo last read() operation. */
+    void undo();
+
+private:
+    /** Reads a single character from the stream. Can be called only in the context of 'read' operation. */
+    QChar get();
```

Comment:
If function has name, starts with `get`, than this function should be `const`.

Reply:
It is not. Same with read()

Reply:
So rename it to `readChar()`

Reply:
My initial idea was that everyone who is familiar with streams will find this name appropriate;
http://www.cplusplus.com/reference/istream/istream/get/

Reply:
OK, but I didn't get it)